SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityUpstreamChecker.java

```diff
@@ -0,0 +1,170 @@
+      upstreamsInBaragonForRequest.removeAll(matches);
+    }
+    return new ArrayList<>(upstreamsInBaragonForRequest);
+  }
+
+  private SingularityLoadBalancerUpdate syncUpstreamsForService(SingularityRequest singularityRequest, SingularityDeploy deploy, Optional<String> loadBalancerUpstreamGroup) throws Exception {
+    final String singularityRequestId = singularityRequest.getId();
+    final LoadBalancerRequestId loadBalancerRequestId = new LoadBalancerRequestId(singularityRequestId, LoadBalancerRequestType.REMOVE, Optional.absent());
+    Collection<UpstreamInfo> upstreamsInBaragonForRequest = lbClient.getLoadBalancerUpstreamsForRequest(loadBalancerRequestId.toString());
+    Collection<UpstreamInfo> upstreamsInSingularityForRequest = getUpstreamsFromActiveTasksForRequest(singularityRequestId, loadBalancerUpstreamGroup);
+    final List<UpstreamInfo> extraUpstreams = getExtraUpstreams(upstreamsInBaragonForRequest, upstreamsInSingularityForRequest);
+    LOG.info("Making and sending load balancer request to remove {} extra upstreams.", extraUpstreams.size());
+    LOG.info("Upstreams removed are: {}", Arrays.toString(extraUpstreams.toArray()));
+    return lbClient.makeAndSendLoadBalancerRequest(loadBalancerRequestId, Collections.emptyList(), extraUpstreams, deploy, singularityRequest);
+  }
+
+  private boolean noPendingDeploy() {
+    return deployManager.getPendingDeploys().size() == 0;
+  }
+
+  public void doSyncUpstreamForService(SingularityRequest singularityRequest) {
+    if (singularityRequest.isLoadBalanced() && noPendingDeploy()) {
+      final String singularityRequestId = singularityRequest.getId();
+      LOG.info("Doing syncing of upstreams for service: {}.", singularityRequestId);
+      final Optional<String> maybeDeployId = deployManager.getInUseDeployId(singularityRequestId);
+      if (maybeDeployId.isPresent()) {
+        final String deployId = maybeDeployId.get();
+        final Optional<SingularityDeploy> maybeDeploy = deployManager.getDeploy(singularityRequestId, deployId);
+        if (maybeDeploy.isPresent()) {
+          final SingularityDeploy deploy = maybeDeploy.get();
+          final Optional<String> loadBalancerUpstreamGroup = deploy.getLoadBalancerUpstreamGroup();
+          final SingularityLoadBalancerUpdate syncUpstreamsUpdate;
+          try {
+            syncUpstreamsUpdate = syncUpstreamsForService(singularityRequest, deploy, loadBalancerUpstreamGroup);
+            final LoadBalancerRequestId loadBalancerRequestId = syncUpstreamsUpdate.getLoadBalancerRequestId();
+            checkSyncUpstreamsState(loadBalancerRequestId, singularityRequestId);
+          } catch (Exception e) {
+            LOG.error("message", e);
+          }
+        }
+      }
+    }
+  }
+
+  private Predicate<BaragonRequestState> baragonRequestStateIsWaiting = baragonRequestState -> baragonRequestState == BaragonRequestState.WAITING;
```

Comment:
for styling purposes, easier to find this if it is at the top of the class and made private static final
```java
private static final Predicate<BaragonRequestState> IS_WAITING_STATE = baragonRequestState -> baragonRequestState == BaragonRequestState.WAITING;
```